scipy/ndimage/tests/test_morphology.py

```diff
@@ -885,104 +1017,136 @@ def test_binary_erosion26(self, dtype):
                     [0, 0, 0, 0, 0, 0, 0, 0],
                     [0, 0, 0, 0, 0, 0, 0, 0],
                     [0, 0, 0, 0, 0, 0, 0, 1]]
-        data = np.array([[0, 0, 0, 0, 0, 0, 0, 0],
-                         [0, 1, 0, 0, 0, 0, 0, 0],
-                         [0, 0, 0, 0, 0, 1, 1, 1],
-                         [0, 0, 1, 1, 1, 0, 1, 1],
-                         [0, 0, 1, 0, 1, 1, 0, 0],
-                         [0, 1, 0, 1, 1, 1, 1, 0],
-                         [0, 1, 1, 0, 0, 1, 1, 0],
-                         [0, 0, 0, 0, 0, 0, 0, 0]], dtype)
+        expected = xp.asarray(expected)
+        data = xp.asarray([[0, 0, 0, 0, 0, 0, 0, 0],
+                           [0, 1, 0, 0, 0, 0, 0, 0],
+                           [0, 0, 0, 0, 0, 1, 1, 1],
+                           [0, 0, 1, 1, 1, 0, 1, 1],
+                           [0, 0, 1, 0, 1, 1, 0, 0],
+                           [0, 1, 0, 1, 1, 1, 1, 0],
+                           [0, 1, 1, 0, 0, 1, 1, 0],
+                           [0, 0, 0, 0, 0, 0, 0, 0]], dtype=dtype)
         out = ndimage.binary_erosion(data, struct, border_value=1,
                                      origin=(-1, -1))
         assert_array_almost_equal(out, expected)
 
-    def test_binary_erosion27(self):
+    def test_binary_erosion27(self, xp):
+        if is_cupy(xp):
+            pytest.xfail("CuPy: NotImplementedError: only brute_force iteration")
+
         struct = [[0, 1, 0],
                   [1, 1, 1],
                   [0, 1, 0]]
+        struct = xp.asarray(struct)
         expected = [[0, 0, 0, 0, 0, 0, 0],
                     [0, 0, 0, 0, 0, 0, 0],
                     [0, 0, 0, 0, 0, 0, 0],
                     [0, 0, 0, 1, 0, 0, 0],
                     [0, 0, 0, 0, 0, 0, 0],
                     [0, 0, 0, 0, 0, 0, 0],
                     [0, 0, 0, 0, 0, 0, 0]]
-        data = np.array([[0, 0, 0, 0, 0, 0, 0],
-                         [0, 0, 0, 1, 0, 0, 0],
-                         [0, 0, 1, 1, 1, 0, 0],
-                         [0, 1, 1, 1, 1, 1, 0],
-                         [0, 0, 1, 1, 1, 0, 0],
-                         [0, 0, 0, 1, 0, 0, 0],
-                         [0, 0, 0, 0, 0, 0, 0]], bool)
+        expected = xp.asarray(expected)
+        data = np.asarray([[0, 0, 0, 0, 0, 0, 0],
+                           [0, 0, 0, 1, 0, 0, 0],
+                           [0, 0, 1, 1, 1, 0, 0],
+                           [0, 1, 1, 1, 1, 1, 0],
+                           [0, 0, 1, 1, 1, 0, 0],
+                           [0, 0, 0, 1, 0, 0, 0],
+                           [0, 0, 0, 0, 0, 0, 0]], dtype=bool)
+        data = xp.asarray(data)
```

Comment:
Here (and elsewhere) the pattern is not just to reduce the diff actually. This is deliberate to work around `xp.bool` being not usable:

```
>>> np.arange(3, dtype=np.bool)
<stdin>:1: FutureWarning: In the future `np.bool` will be defined as the corresponding NumPy scalar.
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/br/miniforge3/envs/scipy-dev/lib/python3.12/site-packages/numpy/__init__.py", line 324, in __getattr__
    raise AttributeError(__former_attrs__[attr])
AttributeError: module 'numpy' has no attribute 'bool'.
`np.bool` was a deprecated alias for the builtin `bool`. To avoid this error in existing code, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:
    https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations. Did you mean: 'bool_'?
>>> np.__version__
'1.26.4'
```

Thus convert to numpy first with `dtype=bool`, then `xp.asarray`.

Reply:
I did understand that - my point was that the indentation changed unnecessarily by going from `np.array` to `np.asarray`. Anyway, not an important critique, just a drive-by observation 😅